sounddevice.py

```diff
@@ -2349,6 +2399,157 @@ def __init__(self, channel_selectors):
+        # through kwargs.
+        del _
+        kind = kwargs['kind']
+        device = kwargs['device']
+        channels = kwargs['channels']
+        # This is to allow some flexibility for other Host-API-specific
+        # settings such as AsioSettings or WasapiSettings and not allow
+        # any accidental reliance on positional parameters.
+
+        if not isinstance(channels, int):
+            raise TypeError('channels must be an int, but found {!r} which is type {!r}'.format(channels, type(channels)))
+
+        if self._channel_selectors!=None and channels!=len(self._channel_selectors):
+            raise ValueError('channels is {}, but the length of channel_selectors is {}'.format(channels, len(self._channel_selectors)))
+
+        this_channelmap = None
+        if self._channel_selectors is not None:
+            # this array must be kept alive!
+            if kind=='input':
+                max_channels = query_devices(device, 'input')['max_input_channels']
+                for ndx in range(len(self._channel_selectors)):
+                    if self._channel_selectors[ndx]>=max_channels:
+                        raise ValueError("channel_selectors[{}]=={} exceeds max_channels=={}".format(ndx, self._channel_selectors[ndx], max_channels))
+                    if self._channel_selectors[ndx]<0:
+                        raise ValueError("channel_selectors[{}]=={} is < 0".format(ndx, self._channel_selectors[ndx]))
+                # No "inversion" is necessary for 'input' devices:
+                this_channelmap = self._channel_selectors
+            elif kind=='output':
+                # "invert" channel_selectors for Core Audio output
+                # devices, as described at
+                # https://lists.apple.com/archives/coreaudio-api/2004/Jan/msg00141.html
+                max_channels = query_devices(device, 'output')['max_output_channels']
+                this_channelmap = [-1 for _ in range(max_channels)]
+                for ndx in range(len(self._channel_selectors)):
+                    if self._channel_selectors[ndx]>=max_channels:
+                        raise ValueError("channel_selectors[{}]=={} exceeds max_channels=={}".format(ndx, self._channel_selectors[ndx], max_channels))
+                    if self._channel_selectors[ndx]<0:
+                        raise ValueError("channel_selectors[{}]=={} is < 0".format(ndx, self._channel_selectors[ndx]))
+                    # This is where we "invert" the channel mapping:
+                    this_channelmap[self._channel_selectors[ndx]]=ndx
+            else:
+                raise ValueError("kind must be either 'input' or 'output', but found {!r} instead".format(kind))
+            #print('Debug: kind={} max_channels={} channel_selectors={!r} this_channelmap={!r}'.format(kind, max_channels, self._channel_selectors, this_channelmap))
+
+        cache_key = kind,device,channels
+        if cache_key in self._cache_cffi_streaminfo:
```

Comment:
Can you clarify why and how you're using this cache? From what I can see this function will never be called more than once, (at Stream instantiation time) therefore the cache will only ever contain one entry.
Or maybe this is a mistake and you mean to declare the cache at the class level (not the instance level)? If so, I'm not sure how much benefit having a cache of MacCoreStreamInfo objects provides. I'm of the mindset that If you're going to add complexity, however minor, it needs to be justified by some objective advantages.